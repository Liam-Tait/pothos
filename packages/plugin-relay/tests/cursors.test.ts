import { execute } from 'graphql';
import { gql } from 'graphql-tag';
import schema from './examples/relay/schema';

describe('cursor based connection', () => {
  it('queries', async () => {
    const query = gql`
      query {
        noArgs: cursorConnection {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        after: cursorConnection(after: 98) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        afterFirst: cursorConnection(after: 95, first: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        afterLast: cursorConnection(after: 95, last: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        afterFull: cursorConnection(after: 40) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        afterFullFirst: cursorConnection(after: 40, first: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        afterLastFull: cursorConnection(after: 40, last: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        before: cursorConnection(before: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        beforeFirst: cursorConnection(before: 5, first: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        beforeLast: cursorConnection(before: 5, last: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        beforeFull: cursorConnection(before: 40) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        beforeFullFirst: cursorConnection(before: 40, first: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
        beforeLastFull: cursorConnection(before: 40, last: 2) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              id
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: {},
    });

    expect(result).toMatchSnapshot();
  });
});
