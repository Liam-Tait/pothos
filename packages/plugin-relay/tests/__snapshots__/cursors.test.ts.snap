// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cursor based connection > queries 1`] = `
{
  "data": {
    "after": {
      "edges": [
        {
          "node": {
            "id": "99",
          },
        },
        {
          "node": {
            "id": "100",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "100",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "99",
      },
    },
    "afterFirst": {
      "edges": [
        {
          "node": {
            "id": "96",
          },
        },
        {
          "node": {
            "id": "97",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "97",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "96",
      },
    },
    "afterFull": {
      "edges": [
        {
          "node": {
            "id": "41",
          },
        },
        {
          "node": {
            "id": "42",
          },
        },
        {
          "node": {
            "id": "43",
          },
        },
        {
          "node": {
            "id": "44",
          },
        },
        {
          "node": {
            "id": "45",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "45",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "41",
      },
    },
    "afterFullFirst": {
      "edges": [
        {
          "node": {
            "id": "41",
          },
        },
        {
          "node": {
            "id": "42",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "42",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "41",
      },
    },
    "afterLast": {
      "edges": [
        {
          "node": {
            "id": "99",
          },
        },
        {
          "node": {
            "id": "100",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "100",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "99",
      },
    },
    "afterLastFull": {
      "edges": [
        {
          "node": {
            "id": "99",
          },
        },
        {
          "node": {
            "id": "100",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "100",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "99",
      },
    },
    "before": {
      "edges": [
        {
          "node": {
            "id": "1",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "1",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "1",
      },
    },
    "beforeFirst": {
      "edges": [
        {
          "node": {
            "id": "1",
          },
        },
        {
          "node": {
            "id": "2",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "2",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "1",
      },
    },
    "beforeFull": {
      "edges": [
        {
          "node": {
            "id": "35",
          },
        },
        {
          "node": {
            "id": "36",
          },
        },
        {
          "node": {
            "id": "37",
          },
        },
        {
          "node": {
            "id": "38",
          },
        },
        {
          "node": {
            "id": "39",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "39",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "35",
      },
    },
    "beforeFullFirst": {
      "edges": [
        {
          "node": {
            "id": "1",
          },
        },
        {
          "node": {
            "id": "2",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "2",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "1",
      },
    },
    "beforeLast": {
      "edges": [
        {
          "node": {
            "id": "3",
          },
        },
        {
          "node": {
            "id": "4",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "4",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "3",
      },
    },
    "beforeLastFull": {
      "edges": [
        {
          "node": {
            "id": "38",
          },
        },
        {
          "node": {
            "id": "39",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "39",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "38",
      },
    },
    "noArgs": {
      "edges": [
        {
          "node": {
            "id": "1",
          },
        },
        {
          "node": {
            "id": "2",
          },
        },
        {
          "node": {
            "id": "3",
          },
        },
        {
          "node": {
            "id": "4",
          },
        },
        {
          "node": {
            "id": "5",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "5",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "1",
      },
    },
  },
}
`;
